experimental/alp/python/alp/benchmark/blas/gemm.py

```diff
@@ -0,0 +1,141 @@
+
+        if trA:
+            A0 = linalg.InitTensorOp([K, M], F32Type.get())
+        else:
+            A0 = linalg.InitTensorOp([M, K], F32Type.get())
+
+        B0 = linalg.InitTensorOp([K, N], F32Type.get())
+        C = linalg.InitTensorOp([M, N], F32Type.get())
+
+        A = linalg.FillOp(output=A0.results[0], value=elem)
+        B = linalg.FillOp(output=B0.results[0], value=elem)
+        std.CallOp(print_pid, [])
+
+        call = std.CallOp(func, [A.results[0], B.results[0], C.results[0]])
+
+        n_iterations = arith.ConstantOp.create_index(niter)
+        start = std.CallOp(rtclock, [])
+        loop = scf.ForOp(zero, n_iterations, one, [])
+        with InsertionPoint(loop.body):
+            call = std.CallOp(func, [A.results[0], B.results[0], C.results[0]])
+            scf.YieldOp([])
+        end = std.CallOp(rtclock, [])
+        treps = arith.SubFOp(end, start)
+        n_iterations_f = arith.ConstantOp(F64Type.get(), float(niter))
+        t = arith.DivFOp(treps, n_iterations_f)
+        flops = arith.DivFOp(nops, t)
+        std.CallOp(print_time, [t.results[0]])
+        std.CallOp(print_flops, [flops.results[0]])
+
+        ret = arith.ConstantOp(IntegerType.get_signless(32), 0)
+        std.ReturnOp(ret)
+
+    return wrapper
+
+
+def generate_benchmark_mlir(func_name, trA, size, reps, dest):
+    with Context() as ctx, Location.unknown() as loc:
+        f32 = F32Type.get()
+        problem_definition = GEMM(trA)
+        mlir_module = Module.create()
+        problem_sizes = {"M": size[0], "N": size[1], "K": size[2]}
+        types = problem_definition.types_mlir_builder(
+            problem_sizes,
+            [f32, f32, f32],
+        )
+        with InsertionPoint(mlir_module.body):
+            gemm = builtin.FuncOp(func_name, (types, [types[-1]]), visibility="private")
+            benchmark = emit_benchmarking_function(trA, size, reps, gemm)
+    save_mlir(str(mlir_module), dest)
+
+
+def main(argv):
+    parser = argparse.ArgumentParser("gemm")
+
+    # Files and paths
+    parser.add_argument("--asm-program", default="")
+    parser.add_argument("--function-name", default="gemm")
+    parser.add_argument("--output", default="")
+
+    # GEMM specific parameters for the problem to benchmark
+    parser.add_argument("--M", type=int)
+    parser.add_argument("--N", type=int)
+    parser.add_argument("--K", type=int)
+    parser.add_argument("--trA", action="store_true")
+
+    # Benchmark specs
+    parser.add_argument("--reps", type=int, default=1)
+    args = parser.parse_args(argv)
+
+    # Generate the benchmark, build and run
+    size = [args.M, args.N, args.K]
+    dest = args.output if args.output else args.function_name + ".bench.mlir"
+    generate_benchmark_mlir(args.function_name, args.trA, size, args.reps, dest)
+
+    if args.asm_program:
+        exe = build(args.asm_program, dest)
+        print(run(exe))
+
+
+if __name__ == "__main__":
```

Comment:
Seems a bit weird that you have to reproduce this file, I'd hope we could reuse.
I'll try to plug in opentuner see if we can reduce the amount of duplication.

Reply:
My tuning experiments have led me to nevergrad, much better !